### ISSUE TITLE
"import ctypes" segfaults on Python 3.6 and Ubuntu 21.10

### ISSUE THREAD

**Issue by d2e3fe16-691a-4f31-a855-a776ce2bd35c** [python/cpython#89911]
BPO | [45748](https://bugs.python.org/issue45748)
--- | :---
Nosy | @ned-deily, @erlend-aasland

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2021-11-09.04:15:20.607>
created_at = <Date 2021-11-08.13:54:44.629>
labels = ['ctypes']
title = '"import ctypes" segfaults on Python 3.6 and Ubuntu 21.10'
updated_at = <Date 2021-11-09.04:15:20.604>
user = 'https://bugs.python.org/ThomasTrummer'
```

bugs.python.org fields:
```python
activity = <Date 2021-11-09.04:15:20.604>
actor = 'ned.deily'
assignee = 'none'
closed = True
closed_date = <Date 2021-11-09.04:15:20.607>
closer = 'ned.deily'
components = ['ctypes']
creation = <Date 2021-11-08.13:54:44.629>
creator = 'Thomas Trummer'
dependencies = []
files = []
hgrepos = []
issue_num = 45748
keywords = []
message_count = 3.0
messages = ['405950', '405952', '405996']
nosy_count = 3.0
nosy_names = ['ned.deily', 'Thomas Trummer', 'erlendaasland']
pr_nums = []
priority = 'normal'
resolution = 'wont fix'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue45748'
versions = ['Python 3.6']
```

</p></details>

**Comment by da4089** [comment]
FWIW, I'm seeing what I think is this problem on Ubuntu 22.04 LTS, using Python 3.6.15, and Ubuntu's libffi 3.4.2-4.

The simple repro is to start the interpreter, and import setuptools.

`(venv) d@ubuntu22-04:~/work/pywinsparkle$ python
Python 3.6.15 (default, Aug  3 2022, 00:05:41) 
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import setuptools
Segmentation fault (core dumped)
(venv) d@ubuntu22-04:~/work/pywinsparkle$`

I haven't yet dug through 3.7+ to see what might have fixed it: if I manage to find a solution, I'll report it here.

**Comment by arhadthedev** [comment]
@da4089 3.7 got its last bugfix release [on June, 2020](https://peps.python.org/pep-0537/#schedule-last-bugfix-release). Nowadays, only security vulnerabilities are fixed; the rest is set in stone.

However, the fix (if applied) will be welcomed on 3.10, the oldest version with bugs being fixed until June, 2023.

**Comment by erlend-aasland** [comment]
> However, the fix (if applied) will be welcomed on 3.10, the oldest version with bugs being fixed until June, 2023.

Quoting https://github.com/python/cpython/issues/89911#issuecomment-1093936920:

> 3.7 and later are all working fine.
>
> First good commit: 55fe1ae9708d81b902b6fe8f6590e2a24b1bd4b0
> First bad commit: fdbd01151dbd5feea3e4c0316d102db3d2a2a412

**Comment by da4089** [comment]
Hi @arhadthedev -- yes, I understand 3.6 is not supported, but some people are required to use it for various reasons, and this issue seemed like the best place to record both confirmations of the issue, and potential fixes or workarounds.  I hope that's ok?

FWIW, I have found a workaround that works for my problem: import ctypes before setuptools.  This seems potentially useful, since it appears to work while not requiring a change to the interpreter (which, as you point out, isn't going to get a new release).

```
d@ubuntu22-04:~$ /opt/PythonVersions/3.6.15/bin/python3
Python 3.6.15 (default, Aug  3 2022, 00:05:41) 
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import setuptools
Segmentation fault (core dumped)
d@ubuntu22-04:~$ /opt/PythonVersions/3.6.15/bin/python3
Python 3.6.15 (default, Aug  3 2022, 00:05:41) 
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> import setuptools
>>> print("w00t -- workaround!")
w00t -- workaround!
>>> 
d@ubuntu22-04:~$ 
```
